File: apps/java/src/main/java/chapter10/DecisionTreeClassifier.java
package chapter10;

import chapter02.Features;
import chapter03.DecisionTree;
import java.util.List;

/** 第 3 章の決定木を変更せずに Classifier に合わせるアダプター。 */
public final class DecisionTreeClassifier implements Classifier {
  private final DecisionTree tree;

  private DecisionTreeClassifier(DecisionTree tree) {
    this.tree = tree;
  }

  /** 深さを制限しない決定木。 */
  public static DecisionTreeClassifier unlimited() {
    return new DecisionTreeClassifier(DecisionTree.unlimited());
  }

  /** 深さの上限を指定した決定木。 */
  public static DecisionTreeClassifier withMaxDepth(int maxDepth) {
    return new DecisionTreeClassifier(DecisionTree.withMaxDepth(maxDepth));
  }

  @Override
  public DecisionTreeClassifier fit(List<Features> x, List<String> t) {
    tree.fit(x, t);
    return this;
  }

  @Override
  public List<String> predict(List<Features> x) {
    return tree.predict(x);
  }
}

File: apps/scala/src/main/scala/machinelearning/chapter10/DecisionTreeClassifier.scala
package machinelearning.chapter10

import machinelearning.chapter02.Features
import machinelearning.chapter03.DecisionTree

/** 第 3 章の決定木を、分類器の約束に合わせるアダプター。第 3 章のコードは変更しない。 */
class DecisionTreeClassifier(tree: DecisionTree) extends Classifier:
  override def fit(x: Vector[Features], t: Vector[String]): DecisionTreeClassifier =
    val _ = tree.fit(x, t)
    this

  override def predict(x: Vector[Features]): Vector[String] = tree.predict(x)

object DecisionTreeClassifier:
  def withMaxDepth(maxDepth: Int): DecisionTreeClassifier =
    DecisionTreeClassifier(DecisionTree.withMaxDepth(maxDepth))

  def unlimited(): DecisionTreeClassifier = DecisionTreeClassifier(DecisionTree.unlimited())
